src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+        checkIndexForAddMethod(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            newNode.next.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            newNode.prev.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> previousToNew = findNodeByIndex(index - 1);
+            newNode.next = previousToNew.next;
+            newNode.prev = previousToNew;
+            previousToNew.next = newNode;
+            previousToNew.next.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        if (findNodeByIndex(index) == null) {
+            return null;
+        }
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexValidation(index);
+        Node<T> necessaryNode = findNodeByIndex(index);
+        T item = necessaryNode.item;
+        necessaryNode.item = value;
+        return item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> current = findNodeByIndex(index);
+        if (head == null) {
+            tail = null;
+        } else if (index == 0) {
+            head = head.next;
+        } else if (index == size - 1) {
+            tail = tail.prev;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = findIndexByValue(object);
+        if (index == -1) {
+            return false;
+        } else {
+            remove(index);
+        }
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexForAddMethod(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("There is no such index in the list, " + index);
+        }
+    }
+
+    private void indexValidation(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("There is no such index in the list, " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
+    }
```

Comment:
You can iterate only first or second half of the list, depending on the index value (also from common mistakes file)